src/deluge/model/action/action_logger.cpp

```diff
@@ -298,6 +300,35 @@ void ActionLogger::recordTempoChange(uint64_t timePerBigBefore, uint64_t timePer
 	}
 }
 
+//record changes to the layout in Performance View so that you can undo changes
+//changes that can be undone:
+//1) loading a new layout
+//2) setting a pad to "hold"
+//3) value editor: changing pad values
+//4) param editor: changing FX assignments to pads
```

Comment:
> Since these are 4 separate actions, maybe we should store them as 4 separate types of consequences? It looks like `ConsequeencePerformanceLayoutChange` is quite large and breaking it in to the 4 separate actions would be both clearer semantically in the code and lighter weight in memory.

They're not entirely separate as they're kind intermingled and work together. Even if I did split them, you'd still need to execute each consequence together in a lot of cases.

So in theory yes it would make sense to split out the consequences. But it would also imply checking exactly what changed or did not change prior to logging a change which is extra cycles and checks every time you're making a change.

I thought it would be lighter to just check if there are any changes instead of specific types of changes and backup everything altogether.

Example 1) Loading a Layout - affects padPress, fxPress, ParamsForPerformance, values
- Loads hold statuses and current hold values
- Loads mappings of values to pads
- Loads mappings of params to columns
- Loads everything associated with each param mapping (paramKind, x, y coordinate, colour)

Example 2) Resetting the view - depending on the context, changes ParamsForPerformance, padPress, fxPress 
- Clears hold statuses and resets values
- Clears param assignments and layout

Example 3) Removing a param from a single column - affects ParamsForPerformance, fxPress, padPress

Reply:
It feels like there should be a more efficient way to do this but I don't have a concrete design in my head at the moment. We can improve this later if it turns out to be trouble.